source/common/secret/sds_api.cc

```diff
@@ -0,0 +1,75 @@
+#include "common/secret/sds_api.h"
+
+#include <unordered_map>
+
+#include "envoy/api/v2/auth/cert.pb.validate.h"
+
+#include "common/config/resources.h"
+#include "common/config/subscription_factory.h"
+#include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
+
+namespace Envoy {
+namespace Secret {
+
+SdsApi::SdsApi(Server::Instance& server, const envoy::api::v2::core::ConfigSource& sds_config,
+               std::string sds_config_name)
+    : server_(server), sds_config_(sds_config), sds_config_name_(sds_config_name), secret_hash_(0) {
+  server_.initManager().registerTarget(*this);
```

Comment:
I don't believe this is correct in all cases. For listener secrets, you are going to need to use the listener init manager, not the server manager. Likewise for cluster secrets you need to use the cluster init manager (which might be the server during startup). You should remove all references to server in these files are pass the init manager directly.

Reply:
Thanks for pointing this out.

I think we can add this method into SdsApi,
```bash
void SdsApi::registerInitTarget(Init::Manager& init_manager) {
  init_manager.registerTarget(*this);
}
```
and add `secret_provider_->registerInitTarget(init_manager)` into `ContextConfigImpl::readCertChainConfig()`.

 We can add accessor `virtual Init::Manager&initManager() PURE` into [Configuration::TransportSocketFactoryContext](https://github.com/envoyproxy/envoy/blob/master/include/envoy/server/transport_socket_config.h#L18).

1. Pass listener init manager to ContextConfigImpl.
As ListenerImpl inherits TransportSocketFactoryContext and provides initManager(), we can modify ServerContextConfigImpl(), and pass context.initManager() into ServerContextConfigImpl() via [DownstreamSslSocketFactory::createTransportSocketFactory()](https://github.com/envoyproxy/envoy/blob/08c139994d97d0513d6f305d17825cd4e74daece/source/extensions/transport_sockets/ssl/config.cc#L39), and then pass initManager to ContextConfigImpl().

1. Pass cluster init manager to ContextConfigImpl.
To pass cluster init manager to ClientContextConfigImpl, we can implement ClusterInfoImpl::initManager(), as ClusterInfoImpl inherits Server::Configuration::TransportSocketFactoryContext(). Then pass initManager to ClientContextConfigImpl() via UpstreamSslSocketFactory::createTransportSocketFactory().

Does this make sense?  I haven't figured out an convenient way to pass initManager to ClusterInfoImpl. Do you have any suggestions? Thanks!

Reply:
No, we definitely cannot add any concept of InitManager to ClusterInfo. The init manager stuff must happen on the main thread only and cluster info is a multi-threaded concept. I would need to sit down with the code in detail to figure out the right way to handle this, but it's possible that the overall design will need rethinking. In the mean time I would consult with @lizan and @PiotrSikora and see if they can help you.

Reply:
@mattklein123 @lizan and @PiotrSikora  I have added a new member sds_init_manager_ into ClusterImplBase, and register SdsApi target at sds_init_manager_. sds_init_manager_ initializes registered targets when ClusterImplBase::onPreInitComplete() is called. 

Both cluster init manager and listener manager is passed to TransportSocketFactoryContext, so that ContextConfigImpl could use init manager to create SdsApi objects.